File: code/mc/mc.c
/*
 *  Implementation of opinion formation model for benchmarking against julia
 *  version. Bottleneck is the RNG.
 */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <omp.h>
#include <boost/random/lagged_fibonacci.hpp> // so much faster than rand() !?


#define n 3         // number of opinions
#define N 900       // number of agents


int main(int argc, char *argv[]){
  if(argc < 2){
    printf("usage: mc p (nruns) (tmax)\n");
    exit(1);
  }
 
  int nruns = 1000;
  double p = atof(argv[1]);
  double tmax = 1e9 * N;
  if(argc > 2){nruns = atoi(argv[2]);}
  if(argc > 3){tmax = atof(argv[3]) * N;}

  double q = 1.0 - p;
  double dN = 1.0 / N;
  srand(time(NULL) + clock());

  double eta = 10;
  p = p * eta / N;
  q = q * eta / N;

  int run = 0;
  
  #pragma omp parallel for
  for(run = 0; run < nruns; ++run){
    boost::lagged_fibonacci607 lf(time(NULL) + clock());
    int s[N] = {0};
    double m[n];
    int k = 0;
    for( ; k < n; ++k)
      m[k] = 1.0 / n;
    int i = 0;
    for( ; i < N; ++i)
      s[i] = i % n;
    double t = 0.0;
    while(t < tmax){
      ++t;
      int a = lf() * N;
      int oa = s[a];
      int neiops[n] = {0};
      int j = 0;
      for( ; j < N; ++j){
        double r = lf();
        if(s[j] == oa){
          if(r < p){
            neiops[s[j]] += 1;
          }        
        }else{
          if(r < q){
            neiops[s[j]] += 1;
          }
        }
      }
      int n_maj = - 1;
      int maj_op = - 1;
      int k = 0;
      for( ; k < n; ++k){
        if(neiops[k] >= n_maj){
          n_maj = neiops[k];
          maj_op = k;
        }
      }

      if(maj_op == oa){
        continue;
      }

      int nmajority = 0;
      k = 0;
      for( ; k < n; ++k){
        if(neiops[k] >= n_maj){
          ++nmajority;
        }
      }
      if(nmajority > 1){
        continue;
      }
      
      m[maj_op] += dN;
      m[oa] -= dN;
      s[a] = maj_op;
      k = 0;
      int done = 0;
      for( ; k < n; ++k){
        if(m[k] > 1.0 - dN){
          done = 1;
        } 
      }
      if(done)
        break;
    }
    printf("%.5f\n",t/N);
    if(int(t) % 100 == 0)
      fflush(stdout);
  }
  return 0;
}

File: code/mc/src/mc.jl
#
# opinion formation 
#
N = 900
dN = 1 / N
tmax = 20
n = 3
p = 0.4
q = 1 - p
eta = 10

normalize(x) = x * eta / N

p = normalize(p)
q = normalize(q)

s = zeros(Int32, N)

for i = 1:n
    s[1 + int((i - 1) * N / n):i * int(N / n)] = i
end


m = zeros(n) + 1/n
mt = zeros(tmax+1,n)
mt[1,:] = m

for i = 1:tmax
    for j = 1:N
        a = randi(N)
        oa = s[a]
        r = rand(N)
        same_op = (s - oa) .== 0
        v = p * same_op + q * !same_op
        neis = r .< v
        neiops = s[neis]
        nop = zeros(n)
        for k = 1:n
            nop[k] = sum(neiops .== k)
        end
        
        maxop_ar = find(nop .>= max(nop))
        
        if length(maxop_ar) > 1
            continue
        else
            maxop = maxop_ar[1]
            m[oa] -= dN
            m[maxop] += dN
            s[a] = maxop
        end
        
    end
        
    mt[i + 1,:] = m
    println(m)

end

for i = 1:n
    plot(mt[:,i])
end
